File: library/src/main/java/io/ashdavies/rx/tasks/SingleTaskOnSubscribe.java
package io.ashdavies.rx.tasks;

import com.google.android.gms.tasks.Task;
import io.reactivex.SingleEmitter;
import io.reactivex.SingleOnSubscribe;

class SingleTaskOnSubscribe<Result> extends TaskOnSubscribe<Result, SingleEmitter<Result>> implements SingleOnSubscribe<Result> {

  SingleTaskOnSubscribe(Task<Result> task) {
    this(task, new SingleTaskListenerFactory<Result>());
  }

  SingleTaskOnSubscribe(Task<Result> task, TaskListenerFactory<Result, SingleEmitter<Result>> factory) {
    super(task, factory);
  }
}

File: library/src/main/kotlin/io/ashdavies/rx/rxtasks/SingleTaskOnSubscribe.kt
package io.ashdavies.rx.rxtasks

import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.android.gms.tasks.Task
import io.reactivex.SingleEmitter
import io.reactivex.SingleOnSubscribe
import java.util.concurrent.Executor

internal class SingleTaskOnSubscribe<T>(private val task: Task<T>, private val executor: Executor) : SingleOnSubscribe<T> {

  override fun subscribe(emitter: SingleEmitter<T>) {
    task.addOnSuccessListener(executor, OnSuccessListener { emitter.onSuccess(it) })
    task.addOnFailureListener(executor, OnFailureListener { emitter.onError(it) })
  }
}
